### PR TITLE
Add EdgeCrafter Model

### LINKED ISSUE #6944: Support EdgeCrafter model for object detection
**Description**
Integrate the EdgeCrafter architecture into the existing Lightning-based training engine to expand the portfolio of supported state-of-the-art edge-optimized models.

**Goals**

- Add EdgeCrafter model implementation to the Lightning engine
- Integrate with the existing training, validation, and inference pipelines
- Support configuration through the current model registry/configuration system
- Enable checkpoint loading and model export where applicable
- Ensure compatibility with the existing dataset preprocessing and augmentation pipeline

**Acceptance Criteria**

- EdgeCrafter is available as a supported model in the Lightning engine
- Users can train, validate, and run inference using the existing API
- Configuration follows the existing Lightning engine conventions
- Documentation includes usage instructions and an example configuration
- Unit and integration tests cover the new model

### PR DESCRIPTION
**Pull request by kprokofi**
### Checklist

- [ ] The PR title and description are clear and descriptive
- [ ] I have manually tested the changes
- [ ] All changes are covered by automated tests
- [ ] All related issues are linked to this PR (if applicable)
- [ ] Documentation has been updated (if applicable)

### CODE REVIEW

**Comment by kprokofi** [comment]
image
image

**Comment by AlbertvanHouten** [comment]
It would be good to include one of the new models in the benchmark_manifest.

**Comment by kprokofi** [comment]
### detection

| Model | Dataset | Scenario | Val f1-score ↑ | Val map ↑ | Val map_50 ↑ | Val map_75 ↑ | Val map_large ↑ | Val map_medium ↑ | Val map_per_class ↑ | Val map_small ↑ | Val mar_1 ↑ | Val mar_10 ↑ | Val mar_100 ↑ | Val mar_100_per_class ↑ | Val mar_large ↑ | Val mar_medium ↑ | Val mar_small ↑ | Train Time ↓ | GPU Mem ↓ | Test Latency ↓ | Status |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| deimv2_l | axial_mri | default | 0.6522 | 0.7292 | 0.9505 | 0.8939 | 0.8465 | 0.7063 | -1.0000 | 0.9000 | 0.6638 | 0.8064 | 0.8638 | -1.0000 | 0.9125 | 0.8600 | 0.9000 | 5m 38s | 6.1 GB | 0.1s | 🆕 |
| deimv2_l | bccd | default | 0.8237 | 0.6502 | 0.9145 | 0.7522 | 0.7476 | 0.5272 | -1.0000 | 0.1965 | 0.4220 | 0.6860 | 0.7870 | -1.0000 | 0.8439 | 0.7299 | 0.3333 | 9m 26s | 6.1 GB | 0.2s | 🆕 |
| deimv2_l | cable_damage | default | 0.9220 | 0.4697 | 0.9351 | 0.4330 | 0.5762 | 0.4551 | -1.0000 | 0.1836 | 0.5138 | 0.5900 | 0.7015 | -1.0000 | 0.7906 | 0.7079 | 0.3800 | 17m 43s | 6.1 GB | 0.1s | 🆕 |
| deimv2_l | vehicles | default | 0.6858 | 0.2969 | 0.4281 | 0.3354 | 0.4527 | 0.3110 | -1.0000 | 0.2689 | 0.3641 | 0.6956 | 0.7259 | -1.0000 | 0.8200 | 0.7282 | 0.5665 | 106m 21s | 6.3 GB | 0.1s | 🆕 |
| edgecrafter_l | axial_mri | default | 0.8913 | 0.7298 | 0.9842 | 0.9177 | 0.7582 | 0.7382 | -1.0000 | 0.8000 | 0.6574 | 0.8213 | 0.8553 | -1.0000 | 0.8938 | 0.8533 | 0.9000 | 7m 20s | 3.2 GB | 0.1s | 🆕 |
| edgecrafter_l | bccd | default | 0.8340 | 0.6572 | 0.9189 | 0.7769 | 0.7534 | 0.5461 | -1.0000 | 0.2139 | 0.4217 | 0.6859 | 0.7775 | -1.0000 | 0.8534 | 0.7204 | 0.3278 | 16m 11s | 3.3 GB | 0.2s | 🆕 |
| edgecrafter_l | cable_damage | default | 0.9267 | 0.4427 | 0.9142 | 0.3882 | 0.5378 | 0.4339 | -1.0000 | 0.2663 | 0.4918 | 0.5904 | 0.6349 | -1.0000 | 0.6534 | 0.6559 | 0.3800 | 39m 58s | 3.2 GB | 0.1s | 🆕 |
| edgecrafter_l | vehicles | default | 0.6901 | 0.3047 | 0.4420 | 0.3384 | 0.4371 | 0.3227 | -1.0000 | 0.2776 | 0.3697 | 0.6982 | 0.7235 | -1.0000 | 0.8224 | 0.7258 | 0.5895 | 146m 27s | 3.4 GB | 0.1s | 🆕 |
| edgecrafter_m | axial_mri | default | 0.7677 | 0.6874 | 0.9576 | 0.8318 | 0.7329 | 0.6848 | -1.0000 | 1.0000 | 0.6596 | 0.8128 | 0.8638 | -1.0000 | 0.8875 | 0.8700 | 1.0000 | 4m 35s | 4.6 GB | 0.1s | 🆕 |
| edgecrafter_m | bccd | default | 0.8312 | 0.6489 | 0.9136 | 0.7687 | 0.7508 | 0.5440 | -1.0000 | 0.1879 | 0.4211 | 0.6855 | 0.7872 | -1.0000 | 0.8483 | 0.7306 | 0.3111 | 10m 51s | 4.7 GB | 0.2s | 🆕 |
| edgecrafter_m | cable_damage | default | 0.9184 | 0.4671 | 0.9291 | 0.3724 | 0.5689 | 0.4533 | -1.0000 | 0.2455 | 0.4964 | 0.5686 | 0.6630 | -1.0000 | 0.7680 | 0.6693 | 0.3800 | 16m 13s | 4.6 GB | 0.1s | 🆕 |
| edgecrafter_m | vehicles | default | 0.6809 | 0.2829 | 0.4180 | 0.3164 | 0.4213 | 0.2954 | -1.0000 | 0.2240 | 0.3532 | 0.6647 | 0.7149 | -1.0000 | 0.8163 | 0.7127 | 0.5658 | 47m 40s | 4.9 GB | 0.1s | 🆕 |
| edgecrafter_s | axial_mri | default | 0.8454 | 0.6790 | 0.9590 | 0.8284 | 0.7289 | 0.6863 | -1.0000 | 0.7333 | 0.6404 | 0.8064 | 0.8915 | -1.0000 | 0.8938 | 0.8933 | 0.9000 | 4m 43s | 3.4 GB | 0.1s | 🆕 |
| edgecrafter_s | bccd | default | 0.8282 | 0.6446 | 0.9075 | 0.7590 | 0.7507 | 0.5226 | -1.0000 | 0.2500 | 0.4202 | 0.6849 | 0.7716 | -1.0000 | 0.8526 | 0.7078 | 0.3111 | 12m 15s | 3.4 GB | 0.2s | 🆕 |
| edgecrafter_s | cable_damage | default | 0.9178 | 0.4445 | 0.9314 | 0.3856 | 0.5662 | 0.4321 | -1.0000 | 0.2137 | 0.4740 | 0.5618 | 0.6479 | -1.0000 | 0.7602 | 0.6444 | 0.3800 | 17m 4s | 3.4 GB | 0.1s | 🆕 |
| edgecrafter_s | vehicles | default | 0.6475 | 0.2199 | 0.3534 | 0.2389 | 0.3915 | 0.2473 | -1.0000 | 0.1943 | 0.3060 | 0.6064 | 0.6363 | -1.0000 | 0.7885 | 0.6358 | 0.4443 | 43m 1s | 3.7 GB | 0.1s | 🆕 |
| edgecrafter_x | axial_mri | default | 0.7387 | 0.6850 | 0.9673 | 0.8974 | 0.7491 | 0.7132 | -1.0000 | 1.0000 | 0.6851 | 0.8362 | 0.8702 | -1.0000 | 0.8938 | 0.8633 | 1.0000 | 11m 47s | 4.1 GB | 0.2s | 🆕 |
| edgecrafter_x | bccd | default | 0.8117 | 0.6426 | 0.9064 | 0.7621 | 0.7446 | 0.5282 | -1.0000 | 0.2095 | 0.4143 | 0.6853 | 0.7734 | -1.0000 | 0.8575 | 0.7112 | 0.2944 | 8m 3s | 4.1 GB | 0.2s | 🆕 |
| edgecrafter_x | cable_damage | default | 0.9103 | 0.4450 | 0.9239 | 0.3873 | 0.5717 | 0.4304 | -1.0000 | 0.1703 | 0.4741 | 0.5774 | 0.6927 | -1.0000 | 0.7947 | 0.6981 | 0.4400 | 23m 58s | 4.0 GB | 0.1s | 🆕 |
| edgecrafter_x | vehicles | default | 0.7000 | 0.3025 | 0.4472 | 0.3391 | 0.4472 | 0.3239 | -1.0000 | 0.2878 | 0.3631 | 0.6931 | 0.7234 | -1.0000 | 0.8207 | 0.7208 | 0.5835 | 145m 28s | 4.2 GB | 0.1s | 🆕 |
| rfdetr_large | axial_mri | default | 0.7350 | 0.7501 | 0.9994 | 0.9457 | 0.8171 | 0.7893 | -1.0000 | 0.8000 | 0.6596 | 0.8489 | 0.8489 | -1.0000 | 0.8813 | 0.8400 | 0.9000 | 8m 45s | 5.3 GB | 0.2s | 🆕 |
| rfdetr_large | bccd | default | 0.8341 | 0.6543 | 0.9212 | 0.7489 | 0.7565 | 0.5326 | -1.0000 | 0.2209 | 0.4179 | 0.6747 | 0.7581 | -1.0000 | 0.8385 | 0.7001 | 0.2722 | 10m 50s | 5.3 GB | 0.2s | 🆕 |
| rfdetr_large | cable_damage | default | 0.9524 | 0.4658 | 0.9546 | 0.4138 | 0.5845 | 0.4501 | -1.0000 | 0.2142 | 0.4991 | 0.5677 | 0.5907 | -1.0000 | 0.6977 | 0.5860 | 0.3000 | 41m 1s | 5.3 GB | 0.1s | 🆕 |
| rfdetr_large | vehicles | default | 0.6865 | 0.2976 | 0.4467 | 0.3343 | 0.4354 | 0.3129 | -1.0000 | 0.2483 | 0.3448 | 0.6709 | 0.7061 | -1.0000 | 0.8065 | 0.7098 | 0.5537 | 89m 43s | 5.5 GB | 0.1s | 🆕 |

File: library/src/getitune/backend/lightning/models/common/edgecrafter_mixin.py [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+        backbone = ECViTAdapter(model_name=cfg[backbone_key], proj_dim=cfg["proj_dim"])
+        encoder = HybridEncoder(model_name=self.model_name)  # type: ignore[attr-defined]
+        decoder = ECTransformer(
+            model_name=self.model_name,  # type: ignore[attr-defined]
+            num_classes=num_classes,
+            eval_spatial_size=input_size,
+            mask_downsample_ratio=4 if with_seg else None,
+        )
+
+        if with_seg:
```
Comment (itallix) [inline library/src/getitune/backend/lightning/models/common/edgecrafter_mixin.py]: I suggest we remove this conditional routing flag from Mixin. Let the Mixin remain completely blind to the concept of "segmentation vs. detection." as it adds some coupling and violates SRP.

Instead, this Mixin can define property stubs and the subclass completes this contract by defining them as standard class variables, e.g.:
```
class EdgeCrafterMixin:
    @property
    def _loss_weights(self) -> dict[str, float]:
        """Return the loss weight dictionary specific to the model task."""
        raise NotImplementedError(
            f"{self.__class__.__name__} must implement the '_loss_weights' property."
        )

    @property
    def _matcher_costs(self) -> dict[str, int | float] | None:
        """Return the matcher cost configuration, or None if matching is not used."""
        raise NotImplementedError(
            f"{self.__class__.__name__} must implement the '_matcher_costs' property."
        )

class EdgeCrafterDetector(EdgeCrafterMixin, LightningDetectionModel):
   _loss_weights = {
        "loss_mal": 1.0,
        "loss_bbox": 5.0,
        "loss_giou": 2.0,
        "loss_fgl": 0.15,
        "loss_ddf": 1.5,
    }
    _matcher_costs = None
```

File: library/src/getitune/backend/lightning/models/common/edgecrafter_mixin.py [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+        ``self.model.optimizer_configuration``.
+
+        Returns:
+            Two lists: optimizer list and lr-scheduler config list.
+        """
+        from getitune.backend.lightning.models.detection.rtdetr import RTDETR
+
+        param_groups = RTDETR._get_optim_params(  # noqa: SLF001
+            self.model.optimizer_configuration,  # type: ignore[attr-defined]
+            self.model,  # type: ignore[attr-defined]
```
Comment (itallix) [inline library/src/getitune/backend/lightning/models/common/edgecrafter_mixin.py]: To avoid using `# type: ignore[attr-defined]`, we can refactor the code by introducing a Protocol for the `self` parameter. This will explicitly define the structure and attributes we expect from any class utilizing this mixin.

File: library/src/getitune/backend/lightning/models/detection/backbones/ecvit.py [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+        self.act = act_layer()
+        self.fc2 = nn.Linear(hidden_features, out_features)
+        self.drop = nn.Dropout(drop)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward pass."""
+        return self.fc2(self.drop(self.act(self.fc1(x))))
+
+
+class Attention(nn.Module):
```
Comment (itallix) [inline library/src/getitune/backend/lightning/models/detection/backbones/ecvit.py]: `Attention`, `Block` and `VisionTransformer` definitions duplicate the code from `library/src/getitune/backend/lightning/models/detection/backbones/vit_tiny.py`

File: library/src/getitune/backend/lightning/models/detection/heads/ec_decoder.py [hunk 4]
```diff
@@ -0,0 +1,1041 @@
...
+        init.constant_(_reg_last.weight, 0)
+
+    def forward(self, scores: Tensor, pred_corners: Tensor) -> Tensor:
+        """Forward pass."""
+        B, L, _ = pred_corners.size()  # noqa: N806
+        prob = F.softmax(pred_corners.reshape(B, L, 4, self.reg_max + 1), dim=-1)
+        prob_topk, _ = prob.topk(self.k, dim=-1)
+        stat = torch.cat([prob_topk, prob_topk.mean(dim=-1, keepdim=True)], dim=-1)
+        quality_score = self.reg_conf(stat.reshape(B, L, -1))
+        return scores + quality_score
```
Comment (itallix) [inline library/src/getitune/backend/lightning/models/detection/heads/ec_decoder.py]: Duplicates definitions from `library/src/getitune/backend/lightning/models/common/layers/transformer_layers.py`

File: library/src/getitune/backend/lightning/models/detection/heads/ec_decoder.py [hunk 5]
```diff
@@ -0,0 +1,1041 @@
...
+
+
+# ---------------------------------------------------------------------------
+# Segmentation Head
```
Comment (itallix) [inline library/src/getitune/backend/lightning/models/detection/heads/ec_decoder.py]: Segmentation specific definitions should go to `segmentation` package

File: library/src/getitune/backend/lightning/models/detection/losses/ec_loss.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+import torch
+import torch.nn.functional as F  # noqa: N812
+from torch import Tensor
+
+from getitune.backend.lightning.models.common.utils.assigners.hungarian_matcher import HungarianMatcher
+
+from .deim_loss import DEIMCriterion
+
+
+class ECCriterion(DEIMCriterion):
```
Comment (itallix) [inline library/src/getitune/backend/lightning/models/detection/losses/ec_loss.py:24]: The loss class should be moved to `common` since it works for both detection and segmentation

File: library/src/getitune/backend/lightning/models/detection/edgecrafter.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+class EdgeCrafter(EdgeCrafterMixin, LightningDetectionModel):  # pyrefly: ignore[inconsistent-inheritance]
+    """getitune Detection model class for EdgeCrafter.
+
+    EdgeCrafter is a family of compact ViT-based DETR models designed for edge
+    deployment via task-specialised distillation.  Four sizes are available:
+    S (small), M (medium), L (large), and X (extra-large).
+
+    Original paper / repository:
+    https://github.com/Intellindust-AI-Lab/EdgeCrafter
```
Comment (itallix) [inline library/src/getitune/backend/lightning/models/detection/edgecrafter.py]:
```suggestion
    Original repository: https://github.com/Intellindust-AI-Lab/EdgeCrafter
    Paper: https://arxiv.org/abs/2603.18739
```

File: library/src/getitune/backend/lightning/models/detection/losses/ec_loss.py (hunk 6 above)
Comment (kprokofi) [reply library/src/getitune/backend/lightning/models/detection/losses/ec_loss.py:24]: I looked into this — common/losses/ currently only holds generic, atomic loss primitives (Focal/CE/GIoU/L1/etc.) that are consumed symmetrically by both detection/ and instance_segmentation/, never a full task-orchestrating Criterion class. Even instance-seg's own criteria (ROICriterion, RPNCriterion, RTMDetInstCriterion) live in instance_segmentation/losses/ rather than common, despite being conceptually cross-cutting too. ECCriterion's architectural siblings (ECViTAdapter, ECDETRDetector, ECTransformer, HybridEncoder) all stay under detection/ as well, despite being consumed by both tasks via EdgeCrafterMixin — moving only the loss would break that consistent grouping. I'd like to keep it in detection/losses/ec_loss.py for now to stay consistent with the rest of the family.

**Comment by kprokofi** [comment]
@itallix  I applied all clean up suggestions in https://github.com/open-edge-platform/geti/pull/6940/commits/bbb56d1060397c722264d5f403acd53c7e9c39a6

Please, review once again

**Comment by kprokofi** [comment]
edge_crafter_l metrics on WGISD after clean up:
image

So, all good here

File: library/src/getitune/backend/lightning/models/detection/heads/ec_decoder.py [hunk 8]
```diff
@@ -0,0 +1,873 @@
...
+            pred_corners_undetach = pred_corners
+            ref_points_detach = inter_ref_bbox.detach()
+            output_detach = output.detach()
+
+        # Segmentation
+        if spatial_feat is not None and self.segmentation_head is not None:
+            dec_out_segs = self.segmentation_head(
+                spatial_features=spatial_feat,
+                query_features=dec_out_hs,
+            )
```
[the remark this answers is not in the document]
Comment (kprokofi) [reply library/src/getitune/backend/lightning/models/detection/heads/ec_decoder.py:317]: Instance Seg is a part of the next PR

**Review by itallix — APPROVED** [review]
